feat(translation): Add dev docs on OCP\Translation (#10098)
nextcloud/documentation: developer_manual/digging_deeper/translation.rst

.. _translation:

===================
Machine Translation
===================

.. versionadded:: 26

Nextcloud offers a **Translation** API. The overall idea is that there is a central OCP API that apps can use to request machine translations of text. To be technology agnostic any app can provide this Translation functionality by registering a Translation provider.

Consuming the Translation API
-----------------------------

> To consume the Translation API, you will need ``\OCP\Translation\ITranslationManager``. This manager offers the following methods:

Reviewer: need to *inject*, I assume? rephraze and link to the DI docs

        }
    }

The method ``getName`` returns a string to identify the registered provider in the user interface.

The method ``translate`` translates the passed string and returns the translation. The two language parameters will be language codes that were returned by ``getAvailableLanguages`` of your provider. In case translation fails, you should throw a ``RuntimeException`` with an explanatory error message.

The class would typically be saved into a file in ``lib/Translation`` of your app but you are free to put it elsewhere as long as it's loadable by Nextcloud's :ref:`dependency injection container<dependency-injection>`.

Providing language detection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

There is also an ``IDetectLanguageProvider`` interface that allows specifying that your provider can auto-detect languages from text input. You can use this as follows:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace OCA\MyApp\Translation;

    use OCA\MyApp\AppInfo\Application;
    use OCP\Files\File;
    use OCP\Translation\ITranslationProvider;
    use OCP\Translation\IDetectLanguageProvider;
    use OCP\IL10N;

>     class Provider implements ITranslationProvider, IDetectLanguageProvider {

Reviewer: nit: highlight this …

        public function __construct(
            private IL10N $l,
        ) {
        }

        public function getName(): string {
            return $this->l->t('My awesome translation provider');
        }

        public function getAvailableLanguages(): array {
            // Return an array of OCP\Translation\LanguageTuple objects here
        }

        public function translate(?string $fromLanguage, string $toLanguage, string $text): string {
            // Do some fancy machine translation and return translated string
        }

        public function detectLanguage(string $text): ?string {
            // Detect the language of $text

>         }

Reviewer: … and these lines to show the additional implementation necessary for the interface

    }

For convenience, when your provider implements language detection, ``null`` will be passed as the first parameter to ``translate``, so you can decide how to most efficiently detect the language as part of the translation process.

If your provider does not implement ``IDetectLanguageProvider``, ``$fromLanguage`` will never be null.

> The method ``detectLanguage`` takes a text in some language and outputs the code of that language, or ``null`` in case detection wasn't successful. The language code that this method returns should match the ones you return in ``getAvailableLanguages``.

Reviewer: ```suggestion
The method ``detectLanguage`` takes a text in some language and outputs the code of that language, or ``null`` in case detection wasn't successful. The language code that this method returns should be one of the languages returned in ``getAvailableLanguages``.

should or has to be?
```

If your provider does not implement ``IDetectLanguageProvider``, ``$fromLanguage`` will never be null.

The method ``detectLanguage`` takes a text in some language and outputs the code of that language, or ``null`` in case detection wasn't successful. The language code that this method returns should be one of the languages returned in ``getAvailableLanguages``.

> should or has to be?

Reviewer: that was a review question :)

Reviewer 2: but it was inside the suggestion :D

Author: oh. right. :see_no_evil: 

It's should. It doesn't have to be.
